### PR TITLE
Provide a @DataMongoTest similar to @DataJpaTest

### PR DESCRIPTION
**Pull request by michael-simons**
Hello everybody,

I like to propose the following slightly new feature for `spring-boot-test-autoconfigure`: This adds a new test slice called `@DataMongoTest` together with an accompanying annotation `@AutoConfigureEmbeddedTestMongod`. `@DataMongoTest` configures an embedded Mongd if `de.flapdoodle.embed:de.flapdoodle.embed.mongo` is on the classpath and enables Spring Data MongoDB (MongoTemplate and MongoRepositories). 

`@AutoConfigureEmbeddedTestMongod` works by conditionally importing `EmbeddedMongoAutoConfiguration` through `EmbeddedTestMongodAutoConfiguration` if `spring.test.mongod.embedded.enabled` is `true` or the annotation is used with `enabled` false. I don't know if that is the correct way, but as there is no embedded Mongod builder like there is for an embedded database, I restrained from writting one. 

`@DataMongoTest` combines the above annotation with `@AutoConfigureDataMongo` which itself uses `MongoAutoConfiguration`, `MongoDataAutoConfiguration` and `MongoRepositoriesAutoConfiguration`.

I had to add two dependencies to test it, see commit 9d6502a3426820cb62d78aa9ead22961f1acceca. Please take note that I had to add an `application.properties` containing `spring.data.mongodb.port = 0` inside the test resources for `spring-boot-test-autoconfigure` to make the embedded Mongod process choose a random port that doesn't conflict with another process fired up through Maven.

I also updated the documentation.

The branched was created from the 1.5.x branch so that `@DataMongTest` may finds it way into Boot with 1.5 as `@JdbcTest` will.

Best regards in hope that this contribution is useful and in a form that you can merge without to much hassle. If it causes problems, I happily take care.

### CODE REVIEW

**Comment by saljuama** [comment]
Nice contribution! although, are you open to extend this PR with some extra feature? such as: 
 - an auto configured TestMongoTemplate in a similar fashion to the TestEntityManager provided in the @DataJpaTest slice. At the moment the most handy utility that i can think of is to load collections from JSON files from the test resources.
 - support for Fongo as well, although this one probably makes sense as a separate enhacement, since the main focus of the PR looks like just providing the slice, but doesn't hurt to ask :)

**Comment by philwebb** [comment]
@saljuama We should take on Fongo as separate task I think.

**Comment by saljuama** [comment]
yea definitely makes sense :)

File: spring-boot-test-autoconfigure/src/main/java/org/springframework/boot/test/autoconfigure/data/mongo/AutoConfigureEmbeddedTestMongod.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ * @since 1.5.0
+ * @see EmbeddedTestMongodAutoConfiguration
+ */
+@Target({ElementType.TYPE, ElementType.METHOD})
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+@Inherited
+@ImportAutoConfiguration
+@PropertyMapping("spring.test.mongod.embedded")
+public @interface AutoConfigureEmbeddedTestMongod {
```
Comment (snicoll) [inline spring-boot-test-autoconfigure/src/main/java/org/springframework/boot/test/autoconfigure/data/mongo/AutoConfigureEmbeddedTestMongod.java]: That sounds like a generic name. One would expect a `AutoconfigureTestMongo` here. Even so, I am not sure what would be its purpose.

What is the use case of having flapdoodle on the classpath and not willing to use it? If that's a thing, maybe we should offer a way for the curent auto-configuration to back-off.

If there isn't a compelling use case, I'd rather remove that altogether

File: spring-boot-test-autoconfigure/src/main/java/org/springframework/boot/test/autoconfigure/data/mongo/EmbeddedTestMongodAutoConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ * Auto-configuration for an embedded mongod process running during tests.
+ *
+ * @author Michael J. Simons
+ * @since 1.5.0
+ * @see AutoConfigureEmbeddedTestMongod
+ */
+@Configuration
+@ConditionalOnProperty(prefix = "spring.test.mongod.embedded", name = "enabled", havingValue = "true", matchIfMissing = true)
+@ImportAutoConfiguration(EmbeddedMongoAutoConfiguration.class)
+public class EmbeddedTestMongodAutoConfiguration {
```
Comment (snicoll) [inline spring-boot-test-autoconfigure/src/main/java/org/springframework/boot/test/autoconfigure/data/mongo/EmbeddedTestMongodAutoConfiguration.java]: Ditto. That "enabled" flag is only meant to conditionally include an auto-configuration. Again, if that's a thing, perhaps there's something we can do in the code to make that more easy.

File: spring-boot-test-autoconfigure/src/main/resources/META-INF/spring.factories [hunk 3]
```diff
@@ -98,3 +98,13 @@ org.springframework.boot.test.autoconfigure.restdocs.RestDocsTestExecutionListen
...
 org.springframework.boot.test.autoconfigure.web.servlet.WebDriverTestExecutionListener
+
+# AutoConfigureEmbeddedTestMongod auto-configuration imports
+org.springframework.boot.test.autoconfigure.data.mongo.AutoConfigureEmbeddedTestMongod=\
+org.springframework.boot.test.autoconfigure.data.mongo.EmbeddedTestMongodAutoConfiguration
+
+# AutoConfigureDataMongo auto-configuration imports
+org.springframework.boot.test.autoconfigure.data.mongo.AutoConfigureDataMongo=\
+org.springframework.boot.autoconfigure.mongo.MongoAutoConfiguration,\
+org.springframework.boot.autoconfigure.data.mongo.MongoDataAutoConfiguration,\
```
Comment (snicoll) [inline spring-boot-test-autoconfigure/src/main/resources/META-INF/spring.factories]: To be clear, I'd advise to include `org.springframework.boot.autoconfigure.mongo.embedded.EmbeddedMongoAutoConfiguration` here.

File: spring-boot-test-autoconfigure/src/test/java/org/springframework/boot/test/autoconfigure/data/mongo/DataMongoTestIntegrationTests.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+	private MongoTemplate mongoTemplate;
+
+	@Autowired
+	private ExampleRepository exampleRepository;
+
+	@Autowired
+	private ApplicationContext applicationContext;
+
+	@Test
+	public void testStuff() {
```
Comment (snicoll) [inline spring-boot-test-autoconfigure/src/test/java/org/springframework/boot/test/autoconfigure/data/mongo/DataMongoTestIntegrationTests.java]: Perhaps you intend to polish that method name? :)

File: spring-boot-test-autoconfigure/src/test/resources/application.properties [hunk 5]
```diff
@@ -0,0 +1,2 @@
+# Use random port for embedded Mongod.
+spring.data.mongodb.port = 0
```
Comment (snicoll) [inline spring-boot-test-autoconfigure/src/test/resources/application.properties]: How is this change related? If that value is sensible why isn't it provided automatically by `@DataMongoTest`?

Comment (michael-simons) [inline spring-boot-test-autoconfigure/src/test/resources/application.properties]: That change isn't related to `@DataMongoTest` but by the port used flapdoodle. Without specifying 0 for random ports, it used the same port for all tests when on the test class path for `spring-boot-test-autoconfigure` which collided with other tests, therefore I set this property. But it's only for the tests of `spring-boot-test-autoconfigure`

Comment (snicoll) [inline spring-boot-test-autoconfigure/src/test/resources/application.properties]: Isn't that a smell? I mean, if that's related to `DataMongoTest` in any way, it should do that automatically, isn't it?

Comment (michael-simons) [inline spring-boot-test-autoconfigure/src/test/resources/application.properties]: That's true. If you didn't already fix this, I will.

File: spring-boot-test-autoconfigure/src/main/java/org/springframework/boot/test/autoconfigure/data/mongo/AutoConfigureEmbeddedTestMongod.java (hunk 1 above)
Comment (michael-simons) [inline spring-boot-test-autoconfigure/src/main/java/org/springframework/boot/test/autoconfigure/data/mongo/AutoConfigureEmbeddedTestMongod.java]: It was meant analogue to the `@DataJpaTest` behavior: If an embedded db is on the classpath, it replaces any explicit or auto-configured database, but that can be turned off. I actually do have DataJpaTests here where use exactly this: One test uses the embedded test database and for the other, I need a specific connection.
I as a user would expect the same choice with Mongo and flapdoodle: Having one test using the embedded mongod, the other not.

File: spring-boot-test-autoconfigure/src/main/resources/META-INF/spring.factories (hunk 3 above)
Comment (michael-simons) [inline spring-boot-test-autoconfigure/src/main/resources/META-INF/spring.factories]: This is something you would include if the embedded Mongod is used unconditionally if flapdoodle is on the path, right?

File: spring-boot-test-autoconfigure/src/main/java/org/springframework/boot/test/autoconfigure/data/mongo/AutoConfigureEmbeddedTestMongod.java (hunk 1 above)
Comment (snicoll) [inline spring-boot-test-autoconfigure/src/main/java/org/springframework/boot/test/autoconfigure/data/mongo/AutoConfigureEmbeddedTestMongod.java]: I am not sure I follow. What you've done is not analogue to `DataJpaTest` IMO. What you've done is selectively adding an auto-configuration, rather than overriding any user's decision.

If you have flapdoodle on the classpath, we'll use that _unless_ you exclude that auto-config. To me it looks like #6809 should do what you want without the need of that machinery.

File: spring-boot-test-autoconfigure/src/main/resources/META-INF/spring.factories (hunk 3 above)
Comment (snicoll) [inline spring-boot-test-autoconfigure/src/main/resources/META-INF/spring.factories]: Sorry, what is the question? 

I'd advise to remove all the machinery around the `enabled` flag and just include that auto-configuration. It looks like you tried to exclude an auto-configuration via a dedicated flag and that's not the right call IMO. Again #6809 seems a better option.

File: spring-boot-test-autoconfigure/src/test/resources/application.properties (hunk 5 above)
Comment (snicoll) [inline spring-boot-test-autoconfigure/src/test/resources/application.properties]: I did not.

File: spring-boot-test-autoconfigure/src/main/java/org/springframework/boot/test/autoconfigure/data/mongo/AutoConfigureEmbeddedTestMongod.java (hunk 1 above)
Comment (michael-simons) [inline spring-boot-test-autoconfigure/src/main/java/org/springframework/boot/test/autoconfigure/data/mongo/AutoConfigureEmbeddedTestMongod.java]: I got your point. Will remove this.

Comment (michael-simons) [inline spring-boot-test-autoconfigure/src/main/java/org/springframework/boot/test/autoconfigure/data/mongo/AutoConfigureEmbeddedTestMongod.java]: After I remove the "machinery" and add the import to spring.factories I suggest, I found no way to exclude the embedded config once flapdoodle is on the path, so #6809 seems important.

What I tried was:

```
@TestConfiguration
@DataMongoTest
@OverrideAutoConfiguration(enabled = true)
@EnableAutoConfiguration(exclude = EmbeddedMongoAutoConfiguration.class)
static class Config {
}
```

Comment (snicoll) [inline spring-boot-test-autoconfigure/src/main/java/org/springframework/boot/test/autoconfigure/data/mongo/AutoConfigureEmbeddedTestMongod.java]: Of course there isn't. That's why #6809 is unresolved 😄

File: spring-boot-test-autoconfigure/src/test/resources/application.properties (hunk 5 above)
Comment (michael-simons) [inline spring-boot-test-autoconfigure/src/test/resources/application.properties]: Would you mind pointing me into the right direction? Would `@TestPropertySource` a good tool?

Comment (snicoll) [inline spring-boot-test-autoconfigure/src/test/resources/application.properties]: No. But I don't know yet how to proceed. Can you join us on gitter when you have some time. We can chat about possible options. Thanks!

Comment (michael-simons) [inline spring-boot-test-autoconfigure/src/test/resources/application.properties]: What about a BeanPostProcessor? Like

```
@Configuration
@EnableConfigurationProperties(MongoProperties.class)
@AutoConfigureBefore(EmbeddedMongoAutoConfiguration.class)
public class TestMongodAutoConfiguration {

	@Bean
	static MongoPropertiesBeanPostProcessor mongoPropertiesBeanPostProcessor() {
		return new MongoPropertiesBeanPostProcessor();
	}

	private static class MongoPropertiesBeanPostProcessor implements BeanPostProcessor {

		@Override
		public Object postProcessBeforeInitialization(Object bean, String beanName) throws BeansException {
			return bean;
		}

		@Override
		public Object postProcessAfterInitialization(Object bean, String beanName) throws BeansException {
			if (bean instanceof MongoProperties && ((MongoProperties) bean).getPort() == null) {
				((MongoProperties) bean).setPort(0);
			}
			return bean;
		}
	}
}
```

To slow? I'll checkout Gitter tomorrow… General Spring Boot room?

**Comment by snicoll** [comment]
#6809 has been fixed. Can you give that a try and see if that works for you?

**Comment by michael-simons** [comment]
I might be doing something wrong. I tried:

```java
@RunWith(SpringRunner.class)
@DataMongoTest
public class DataMongoTestDisableEmbeddedMongoTest {

	@Autowired
	private ApplicationContext context;

	@Test
	public void testContextLoads() throws Exception {

		assertThat(this.context).isNotNull();
		assertThat(this.context.getBeanNamesForType(MongodExecutable.class)).isEmpty();
	}

	@TestConfiguration
	@ImportAutoConfiguration(exclude = EmbeddedMongoAutoConfiguration.class)
	static class Config {

	}
}
```

With `@ImportAutoConfiguration(exclude = EmbeddedMongoAutoConfiguration.class)` expected to not include the embedded mongo auto configuration if it is specified by `DataMongoTest`. Result is:

```
java.lang.IllegalStateException: The following classes could not be excluded because they are not auto-configuration classes:
	- org.springframework.boot.autoconfigure.mongo.embedded.EmbeddedMongoAutoConfiguration

	at org.springframework.boot.autoconfigure.EnableAutoConfigurationImportSelector.checkExcludedClasses(EnableAutoConfigurationImportSelector.java:172)
	at org.springframework.boot.autoconfigure.EnableAutoConfigurationImportSelector.selectImports(EnableAutoConfigurationImportSelector.java:89)
	at org.springframework.context.annotation.ConfigurationClassParser.processDeferredImportSelectors(ConfigurationClassParser.java:474)
	at org.springframework.context.annotation.ConfigurationClassParser.parse(ConfigurationClassParser.java:184)
	at org.springframework.context.annotation.ConfigurationClassPostProcessor.processConfigBeanDefinitions(ConfigurationClassPostProcessor.java:308)
	at org.springframework.context.annotation.ConfigurationClassPostProcessor.postProcessBeanDefinitionRegistry(ConfigurationClassPostProcessor.java:228)
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanDefinitionRegistryPostProcessors(PostProcessorRegistrationDelegate.java:270)
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:93)
	at org.springframework.context.support.AbstractApplicationContext.invokeBeanFactoryPostProcessors(AbstractApplicationContext.java:686)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:524)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:736)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:369)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:313)
	at org.springframework.boot.test.context.SpringBootContextLoader.loadContext(SpringBootContextLoader.java:118)
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContextInternal(DefaultCacheAwareContextLoaderDelegate.java:98)
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContext(DefaultCacheAwareContextLoaderDelegate.java:116)
	at org.springframework.test.context.support.DefaultTestContext.getApplicationContext(DefaultTestContext.java:83)
	at org.springframework.boot.test.autoconfigure.SpringBootDependencyInjectionTestExecutionListener.prepareTestInstance(SpringBootDependencyInjectionTestExecutionListener.java:47)
	at org.springframework.test.context.TestContextManager.prepareTestInstance(TestContextManager.java:230)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.createTest(SpringJUnit4ClassRunner.java:228)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner$1.runReflectiveCall(SpringJUnit4ClassRunner.java:287)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.methodBlock(SpringJUnit4ClassRunner.java:289)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:247)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:94)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:70)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:191)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:283)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:173)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:128)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:203)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:155)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:103)
```

**Comment by snicoll** [comment]
Totally unrelated to this PR, but yeah, looks weird.

**Comment by michael-simons** [comment]
Just tried to do what you asked for :) Tested #6809 with my use case…

**Comment by snicoll** [comment]
Look at the commit, there is an exclude attribute there that can be added on the annotation. Maybe you could try that?

As for that stacktrace, I would have reported that on the original ticket.

**Comment by michael-simons** [comment]
Using the alias on "my" annotation works as expected…

**Comment by michael-simons** [comment]
Any opinion on that https://github.com/spring-projects/spring-boot/pull/7600#discussion_r94067969 suggestion for setting `spring.data.mongodb.port` to random if not defined? As far as I see, the least invasive option for the user.

File: spring-boot-test-autoconfigure/src/test/resources/application.properties (hunk 5 above)
Comment (snicoll) [inline spring-boot-test-autoconfigure/src/test/resources/application.properties]: I wish there was another way. I'll give it more thoughts.

**Comment by saljuama** [comment]
@michael-simons been taking a look at your comment, and in regards of possible side effects, MongoProperties has a method to create a MongoClient, is there any risk of the MongoClient being created before the post processing happens? Probably its a silly question, as I'm not that familiar with BeanPostProcessors, my apologies if that is the case. And I'm not aware of any other mechanism that could serve the purpose and it is not invasive. 

Also, in regards of https://github.com/spring-projects/spring-boot/pull/7600#issuecomment-268639732 would you prefeer that i open a PR to your repository, or better that i wait for the PR to be closed and then i open a new one afterwards?

**Comment by michael-simons** [comment]
Hello @saljuama 
I'm not a big fan of overriding core classes… The post processor is instantiated before the MongoProperties which are instantiated before the client. That works just fine, but could maybe have a performance penalty with a lot of beans.

Regarding the TestMongoTemplate I would wait for this to be merged or not.

**Comment by snicoll** [comment]
I am not a huge fan of the current proposal but I have no idea how to fix it differently. I need to resume my review of this PR and we can then iterate after that.

**Comment by saljuama** [comment]
@michael-simons fair enough :+1:

**Comment by snicoll** [comment]
Alright, I've polished this PR on my fork and I was about to push it when I noticed that the tests were breaking now. If we have the embedded mongodb thing, it starts on a fixed port by default so as soon as you have more than one integration tests in the same module, it blows up because the port is already in use.

I initially thought that this problem was `DataMongoTest` specific so I've [managed to override the port to 0](https://github.com/snicoll/spring-boot/blob/df33349a93d2cb9ec693cd36e10b3403f4f0288e/spring-boot-test-autoconfigure/src/main/java/org/springframework/boot/test/autoconfigure/data/mongo/DataMongoTestContextBootstrapper.java#L42-L44) with the limitation that if you want to use a real MongoDB server, [you'll need to set the port yourself](https://github.com/snicoll/spring-boot/blob/df33349a93d2cb9ec693cd36e10b3403f4f0288e/spring-boot-test-autoconfigure/src/test/java/org/springframework/boot/test/autoconfigure/data/mongo/DataMongoTestDisableEmbeddedMongoTest.java#L40-L41). This sounds quite ok in the context of `DataMongoTest` but it isn't at all in a broader context.

So what now? It looks like what we have is quite good actually and users of boot with the embedded mongo support probably have the port set to 0 already. 

Any insight?

**Comment by michael-simons** [comment]
I know that the new dependency to embedded mongod blows up the tests, that's why I modified / added application.properties with `spring.data.mongodb.port = 0` in the first place.

I like your `DataMongoTestContextBootstrapper`. "My" bean post processor is only active during a `DataMongoTest` so the other integration tests would still break. Does the bootstrapper prevent that? 

If so, I'd say that this is fine the way it is. We have set the embedded mongo port to 0 anyway, and I assume others would do the same…

**Comment by snicoll** [comment]
> Does the bootstrapper prevent that?

It does not.

**Comment by snicoll** [comment]
Alright, I had another pass with @philwebb and here is our suggestion.

Create a `spring.mongodb.embedded.port` that defaults to `27017` for backward compatible reason. The new `DataMongoTest` sets that property to 0 automatically but also gives you a `randomPort` boolean property (`true` by default) that you can switch to false to run on a given port. 

```
@DataMongoTest(randomPort = false)
@TestPropertySource(properties = "spring.data.mongodb.embedded.port=12345")
```

Of  course, if you are not using the embedded support, you are not affected by this change. The client will connect to the configured port as before as there is no override now.

The current `spring.mongodb.port` can't use a value of 0 now (it will throw an exception) and is only used for the regular client/server setup. This has the advantage to remove this weird semantic that 0 is only a valid option when the embedded mode is used.

@michael-simons @wilkinsona let me know if we're missing anything there.

**Comment by wilkinsona** [comment]
What happens when `spring.mongo.port` and `spring.mongo.embedded.port` are both set? Which one will win and how will the auto-configured client get configure with the "right" one?

**Comment by snicoll** [comment]
Alright, we're not really sure we fully understand how this is going to be used. And the side effects are going to be quite tricky.

The use of `@DataMongoTest` is going to create a context, potentially with an embedded mongodb instance it it. If you are using slicing, you are probably using also `@SpringBootTest`  which is going to create another context (and another embedded mongo instance).

We basically have two choices:

1. Do nothing in `@DataMongoTest`. If you've defined `spring.data.mongodb.port=0` in your config somewhere, we're going to happily create different mongo instances for you. If you haven't, the test will fail attempting to create another instance. You're probably have that issue now with different config for `@SpringBootTest`
2. Try to be smart and automatically set the port to `0` when you're using `@DataMongoTest`. The problem is that we don't know if the embedded support is active so it can lead to all kinds of issues (including overriding a value you've set yourself).

Can you please describe how you'd be using this new annotations?

**Comment by michael-simons** [comment]
I use this to annotate one plain test (not a `@SpringBootTest`) with `@DataMongoTest` to get: A mongo client and if available, Spring Data Mongo repositories and nothing else. `@DataMongoTest` should also provide an embedded mongo db for those tests.
I am happy to explicitly provide `spring.data.mongodb.port=0` myself because I have to add the dependency myself and assume that I know what am doing. But I want to have `@DataMongoTest` enable it.

I don't expect Spring Boot to solve that problem of random ports for me at that point.

**Comment by snicoll** [comment]
Alright so in the meantime  we've implemented #8044 which does use an random port for embedded mongodb if there is no configuration. That way this provides a sensible behaviour for both `@DataMongoTest` and regular use cases as well.

Thanks @michael-simons for the PR and the discussion!

**Comment by philwebb** [comment]
@michael-simons I hope you don't mind, but I've updated the `@author` tags from `Michael J. Simons` to `Michael Simons`. The initial looks odd next to all the other tags. If you want a longer name in tag, let me know your middle name and I'll update them.

Thanks again for the PR!

**Comment by michael-simons** [comment]
😂 absolutely fine with me.
